File: src/jastudio/jastudio/note/notefields/tag_flag_field.py
from __future__ import annotations

from typing import TYPE_CHECKING, override

from autoslot import Slots  # pyright: ignore[reportMissingTypeStubs]

if TYPE_CHECKING:
    from jastudio.note.jpnote import JPNote
    from jastudio.note.tag import Tag
    from sysutils.weak_ref import WeakRef

class TagFlagField(Slots):
    def __init__(self, note: WeakRef[JPNote], tag: Tag) -> None:
        self._note: WeakRef[JPNote] = note
        self.tag: Tag = tag
        # Cache tag state during construction - invalidated when matching_configuration is recreated
        self._cached_is_set: bool = note().tags.contains(tag)

    def is_set(self) -> bool:
        return self._cached_is_set

    def set_to(self, set_: bool) -> None:
        if set_:
            self._note().tags.set(self.tag)
        else:
            self._note().tags.unset(self.tag)

    @override
    def __repr__(self) -> str: return f"""{self.tag.name}: {self.is_set()}"""

File: src/src_dotnet/JAStudio.Core/Note/NoteFields/TagFlagField.cs
namespace JAStudio.Core.Note.NoteFields;

public class TagFlagField
{
    private readonly JPNote _note;
    public readonly Tag Tag;
    private readonly bool _cachedIsSet;

    public TagFlagField(JPNote note, Tag tag)
    {
        _note = note;
        Tag = tag;
        // Cache tag state during construction - invalidated when matching_configuration is recreated
        _cachedIsSet = note.Tags.Contains(tag);
    }

    public bool IsSet()
    {
        return _cachedIsSet;
    }

    public void SetTo(bool set)
    {
        if (set)
        {
            _note.Tags.Set(Tag);
        }
        else
        {
            _note.Tags.Unset(Tag);
        }
    }

    public override string ToString()
    {
        return $"{Tag.Name}: {IsSet()}";
    }
}
